manta-trusted-setup/src/groth16/ceremony/mod.rs

```diff
@@ -16,39 +16,135 @@
 
 //! Groth16 Trusted Setup Ceremony
 
-use crate::mpc;
+use crate::{
+    ceremony::{
+        participant::{Participant, Priority},
+        signature::SignatureScheme,
+    },
+    groth16::mpc::{Configuration, State},
+};
+use alloc::string::String;
+use core::fmt::Debug;
+use manta_util::{
+    collections::vec_deque::MultiVecDeque,
+    serde::{Deserialize, Serialize},
+};
 
-pub mod registry;
+pub mod client;
+pub mod config;
+pub mod message;
 
 #[cfg(feature = "std")]
-#[cfg_attr(doc_cfg, doc(cfg(feature = "std")))]
 pub mod coordinator;
 
-#[cfg(all(feature = "bincode", feature = "serde"))]
-#[cfg_attr(doc_cfg, doc(cfg(all(feature = "bincode", feature = "serde"))))]
-pub mod signature;
-
-/// Participant
-pub trait Participant {
-    /// Participant Identifier Type
-    type Identifier;
-
-    /// Returns the [`Identifier`](Self::Identifier) for `self`.
-    fn id(&self) -> &Self::Identifier;
+#[cfg(feature = "std")]
+pub mod server;
 
-    /// Returns the priority level for `self`.
-    ///
-    /// # Note
-    ///
-    /// Lower level indicates a higher priority.
-    fn level(&self) -> usize;
-}
+/// Participant Queue Type
+pub type Queue<C, const LEVEL_COUNT: usize> =
+    MultiVecDeque<<C as Ceremony>::Identifier, LEVEL_COUNT>;
 
 /// Ceremony Configuration
-pub trait Ceremony: mpc::Types {
+pub trait Ceremony: Configuration + SignatureScheme {
     /// Participant Identifier Type
     type Identifier: Clone + PartialEq;
 
+    /// Participant Priority Type
+    type Priority: Into<usize>;
+
     /// Participant Type
-    type Participant: Participant<Identifier = Self::Identifier>;
+    type Participant: Participant<
+            Identifier = Self::Identifier,
+            VerifyingKey = Self::VerifyingKey,
+            Nonce = Self::Nonce,
+        > + Priority<Priority = Self::Priority>;
+}
+
+/// MPC States
+#[derive(Deserialize, Serialize)]
+#[serde(
+    bound(
+        deserialize = "C::Challenge: Deserialize<'de>",
+        serialize = "C::Challenge: Serialize",
+    ),
+    crate = "manta_util::serde",
+    deny_unknown_fields
+)]
+pub struct MpcState<C>
```

Comment:
It is unclear whether this is a `Vec` of `State`s and `Challenge`s because it represents the history of a single ceremony or because it represents the current state of a ceremony that involves multiple circuits.  In other words, would the indices of this vector refer to individual contributions or individual circuits?

From context I can tell it's the latter, i.e. that the `state` field contains a `Vec` of `ProverKey`s for multiple circuits.  Let's clarify in the documentation.

Reply:
Using the new `Parallel` API is this clearer?

Reply:
Yes, much clearer, thanks!